### PR TITLE
Check the integrity of the released source tarball

### PR DESCRIPTION
**Pull request by reneme**
... if we build documents for a released version (i.e. `$BOTAN_VERSION=x.y.z`) we download the .tar.xz file from Botan's upstream website (https://botan.randombit.net) and compare its contents to the result of a fresh git checkout. The content should be equal (except for version.txt).

If the build is targetting a development version (that has no signed and released upstream tarball), the check is simply skipped.

This is how the CI looks when building a release version:

image

(I forgot to take a USB-C to USB-A adapter, and couldn't use my YubiKey to sign those commits)

### CODE REVIEW

File: tools/tarball_checker/check.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+def recursive_diff(dir1, dir2):
+    compared = dircmp(dir1, dir2)
+    differing_paths = []
+    if (compared.left_only or compared.right_only or compared.diff_files
+        or compared.funny_files):
+        return compared.left_only + compared.right_only + compared.diff_files + compared.funny_files
+    for subdir in compared.common_dirs:
+        if diff_entries := recursive_diff(os.path.join(dir1, subdir), os.path.join(dir2, subdir)):
+            differing_paths += [os.path.join(subdir, entry) for entry in diff_entries]
```
Comment (FAlbertDev) [inline tools/tarball_checker/check.py:66]: Don't we need to check that there aren't more (non-common) subdirectories than `compared.common_dirs` in either directory?

**Review by FAlbertDev — APPROVED** [review]
Jia Tan hates this trick.

File: tools/tarball_checker/check.py (hunk 1 above)
Comment (reneme) [reply tools/tarball_checker/check.py:66]: The way I understand [the documentation](https://docs.python.org/3/library/filecmp.html), those show up in `left_only` and `right_only` and would therefore fail the comparison as is.

Comment (reneme) [reply tools/tarball_checker/check.py:66]: Please merge, when you are convinced by this. :)

Comment (FAlbertDev) [reply tools/tarball_checker/check.py:66]: Yeah, I think you are right!

**Pull request merged**